libs/langchain-azure-cosmosdb/src/caches.ts

```diff
@@ -0,0 +1,171 @@
+import { Document } from "@langchain/core/documents";
+import { EmbeddingsInterface } from "@langchain/core/embeddings";
+import { CosmosClient, CosmosClientOptions } from "@azure/cosmos";
+import { DefaultAzureCredential } from "@azure/identity";
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import {
+  AzureCosmosDBNoSQLConfig,
+  AzureCosmosDBNoSQLVectorStore,
+} from "./azure_cosmosdb_nosql.js";
+
+const USER_AGENT_SUFFIX = "LangChain-CDBNoSQL-SemanticCache-JavaScript";
+
+// Create a new object based on dbConfig, and modify the 'client' property with user agent.
+function updateDbConfig(
+  dbConfig: AzureCosmosDBNoSQLConfig,
+  client: CosmosClient
+) {
+  const updatedDbConfig = {
+    ...dbConfig,
+    client,
+  };
+
+  return updatedDbConfig;
+}
+
+/**
+ * Represents a Semantic Cache that uses CosmosDB NoSQL backend as the underlying
+ * storage system.
+ *
+ * @example
+ * ```typescript
+ * const embeddings = new OpenAIEmbeddings();
+ * const cache = new AzureCosmosDBNoSQLSemanticCache(embeddings, {
+ *   databaseName: DATABASE_NAME,
+ *   containerName: CONTAINER_NAME
+ * });
+ * const model = new ChatOpenAI({cache});
+ *
+ * // Invoke the model to perform an action
+ * const response = await model.invoke("Do something random!");
+ * console.log(response);
+ * ```
+ */
+export class AzureCosmosDBNoSQLSemanticCache extends BaseCache {
+  private embeddings: EmbeddingsInterface;
+
+  private config: AzureCosmosDBNoSQLConfig;
+
+  private cacheDict: { [key: string]: AzureCosmosDBNoSQLVectorStore } = {};
+
+  constructor(
+    embeddings: EmbeddingsInterface,
+    dbConfig: AzureCosmosDBNoSQLConfig
+  ) {
+    super();
+    let client: CosmosClient;
+
+    const connectionString =
+      dbConfig.connectionString ??
+      getEnvironmentVariable("AZURE_COSMOSDB_NOSQL_CONNECTION_STRING");
+
+    const endpoint =
+      dbConfig.endpoint ??
+      getEnvironmentVariable("AZURE_COSMOSDB_NOSQL_ENDPOINT");
+
+    if (!dbConfig.client) {
+      if (connectionString) {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        let [endpoint, key] = connectionString!.split(";");
+        [, endpoint] = endpoint.split("=");
+        [, key] = key.split("=");
+
+        client = new CosmosClient({
+          endpoint,
+          key,
+          userAgentSuffix: USER_AGENT_SUFFIX,
+        });
+      } else {
+        // Use managed identity
+        client = new CosmosClient({
+          endpoint,
+          aadCredentials: dbConfig.credentials ?? new DefaultAzureCredential(),
+          userAgentSuffix: USER_AGENT_SUFFIX,
+        } as CosmosClientOptions);
+      }
+      this.config = updateDbConfig(dbConfig, client);
+    } else {
+      this.config = dbConfig;
+    }
+    this.embeddings = embeddings;
+  }
+
+  private getLlmCache(llmKey: string) {
+    const key = getCacheKey(llmKey);
+    if (!this.cacheDict[key]) {
+      this.cacheDict[key] = new AzureCosmosDBNoSQLVectorStore(
```

Comment:
In the current state of this implementation, this use the same default container name as for the VectorStore, which can be problematic:

For example, if a user uses default values and have a vector store for RAG and semantic cache, the results will get mixed up.

I suggest 2 changes:
- Add a metadata to indicate that the document is used for semantic caching (so users wants to use a single table, they can use filters to distinguish cache vs documents)
- Use a different container name for the semantic caching by default, for example `vectorSearchContainer` to avoid conflicts

Reply:
added different default database and container name for semantic cache